torch/_logging/internal.py

```diff
@@ -0,0 +1,398 @@
+        else:
+            raise ValueError(
+                f"Invalid log settings: {settings}, must be a comma separated list of log or artifact names."
+            )
+
+    return log_state
+
+def _is_valid_module(qname):
+    try:
+        __import__(qname)
+        return True
+    except ImportError:
+        return False
+
+
+class FilterByType(logging.Filter):
+    def __init__(self, enabled_types):
+        self.set_enabled_types(enabled_types)
+
+    def filter(self, record):
+        return isinstance(record.msg, self.enabled_types)
+
+    def set_enabled_types(self, enabled_types):
+        self.enabled_types = tuple(set(enabled_types))
+
+
+def _get_log_state():
+    log_setting = os.environ.get("TORCH_LOGS", None)
+    if log_setting is None:
+        return log_state
+    else:
+        return _parse_log_settings(log_setting)
+
+
+# setup custom handlers
+# if the log level of a component is set to INFO, setup
+# an additional handler to print those messages, because
+# the debug handler is what handles custom objects like guards,
+# bytecode, etc.
+# if the log level of a component is set to DEBUG, allow all
+# string messages and allowed types (other than those off by default)
+def _setup_handlers(create_handler_fn, log, enabled_types, level=None):
+    debug_handler = create_handler_fn()
+    debug_handler.setFormatter(DEFAULT_FORMATTER)
+    debug_handler.setLevel(logging.DEBUG)
+
+    if level == logging.DEBUG:
+        enabled_types = enabled_types.union({str})
+
+    filter = FilterByType(enabled_types)
+    debug_handler.addFilter(filter)
+    log.addHandler(debug_handler)
+
+    if level is not None and level > logging.DEBUG:
+        generic_handler = create_handler_fn()
+        generic_handler.setFormatter(DEFAULT_FORMATTER)
+        generic_handler.setLevel(level)
+        log.addHandler(generic_handler)
+
+
+# mark handlers that we've created
+# so we don't modify user handlers
+def _tag_handler(handler):
+    handler.__torch_log_handler = True
+    return handler
+
+
+def _is_torch_handler(handler):
+    return hasattr(handler, "__torch_log_handler")
+
+
+# clears all torch handlers on specified loggers
+def _clear_handlers(log):
+    to_remove = [handler for handler in log.handlers if _is_torch_handler(handler)]
+    for handler in to_remove:
+        log.removeHandler(handler)
+
+def _get_registered_parent_qname(log_qname):
+    logger = logging.getLogger(log_qname)
+    while logger.parent:
+        if log_registry.is_log(logger.parent):
+            return logger.parent.name
+        logger = logger.parent
+
+    return None
+
+# initialize loggers log_names
+# each developer component should call this for their own logs
+# in the appropriate location after relevant types have been registered
+def _init_logs(log_file_name=None):
+    for log_name in log_registry.get_log_qnames():
+        log = logging.getLogger(log_name)
+        log.setLevel(logging.DEBUG)  # allow all messages through to the handlers
```

Comment:
I'm not sure about that. From the discussion in the RFC about logging. There seemed to be agreement that we should not log anything by default unless the user asks for logging. Should this be replaced to be CRITICAL + 1 or soemthing?

Reply:
The console handler has a filter for what is seen by the user - nothing is actually getting logged here

Reply:
It's following this pattern: https://docs.python.org/3/howto/logging-cookbook.html#logging-cookbook

Reply:
Whatever the resolution of this conversation is, it needs to be clearly explained in comments here

Reply:
I read through the cookbook and I did not interpret it the same way you did. Also, this code still seems inconsistent with the discussion we had on Monday on VC. Warning, long essay coming through here.

I think there are two users of the logging API: applications and libraries. An application has global control over all the components in its system and is capable of making decisions, e.g., about where it wants logs to go, etc. A library doesn't really know what the end user wants to do with the logs: should they go to stderr? A file? Etc. So the library's job is to put logs in the conventional place, and then let the user make the final call about what to do with logs. For this reason, there are multiple admonitions in the logging cookbook along these lines:

> Note It is strongly advised that you do not add any handlers other than [NullHandler](https://docs.python.org/3/library/logging.handlers.html#logging.NullHandler) to your library’s loggers. This is because the configuration of handlers is the prerogative of the application developer who uses your library. The application developer knows their target audience and what handlers are most appropriate for their application: if you add handlers ‘under the hood’, you might well interfere with their ability to carry out unit tests and deliver logs which suit their requirements. -- https://docs.python.org/3/howto/logging.html#configuring-logging-for-a-library

So, are we an application or a library? PyTorch is clearly a library, but it is *also* in the business of providing an **optional** application level logging system, that a user can use if they do not otherwise have an application-wise logging setup (which is the common case for quick scripts that are just using the PyTorch API). The keyword here is optional: a user may choose to not use any of PyTorch's affordances and go its own route.

OK, so returning to the question at hand. Should we set the logLevel on the loggers defined by our library? https://docs.python.org/3/howto/logging.html#configuring-logging-for-a-library clearly says that the expected default is only logs of WARNING level or higher should be printed to the user. So it is wrong to set the logLevel to DEBUG. The Logging Cookbook you have quoted is irrelevant, because this cookbook is talking about how an *application* should configure logs. In this case, the application knows one of its handlers wants debug logs, so it configures the logger to emit debug logs. Seems reasonable, and also seems inapplicable for the design problem here.

You have argued, however, that it is OK for us to set the log level to DEBUG, because we install handlers which then filter out DEBUG log messages. But I argued above, PyTorch's installed handlers are OPTIONAL. A user may choose not to use them and do their own thing. By default, we ought not deluge their handlers with DEBUG logs that they didn't ask for.

----

If you are still not convinced, let me present two more arguments.

Extra Argument 1: One thing I pointed out in VC is that if you set the logger's level to DEBUG, then isEnabledFor calls will incorrectly report the logger is enabled for DEBUG, even if there isn't actually any handler that wants DEBUG logs. You argued that this is an antipattern and people shouldn't make isEnabledFor calls to check if we do log messages or not. However, the Python documentation *explicitly* suggests this as a thing you might want to do, if constructing log arguments is expensive:

> Formatting of message arguments is deferred until it cannot be avoided. However, computing the arguments passed to the logging method can also be expensive, and you may want to avoid doing it if the logger will just throw away your event. To decide what to do, you can call the [isEnabledFor()](https://docs.python.org/3/library/logging.html#logging.Logger.isEnabledFor) method which takes a level argument and returns true if the event would be created by the Logger for that level of call. --  https://docs.python.org/3/howto/logging.html#optimization

We can overrule standard library usage guidance, but we better have a good reason for it, and I haven't seen a good reason yet.

Extra Argument 2: I don't think you even need this, from a technical perspective. If I enable more verbose logs for Dynamo, my application would quite logically set the log level on the Dynamo logger. Do it once you actually configure the verbosity!

Reply:
FWIW, I second Edward's suggestion that the default loglevel should be `WARNING`

Reply:
Chatted with Ed in vc, going to switch this to use dedicated loggers for artifacts instead, which should simplify the design further

Reply:
The level issue is resolved in the latest iteration, please review

Reply:
I also addressed the concerns in Ed's essay by actually using the hierarchical part of hierarchical logging :)